libs/constrained-generators/testlib/Test/Min/Base.hs

```diff
@@ -0,0 +1,501 @@
+import qualified Data.List.NonEmpty as NE
+import Data.String (fromString)
+import Data.Typeable
+import GHC.Stack
+import Prettyprinter
+
+-- ===========================================
+-- Terms
+-- ===========================================
+
+type AppRequires t dom rng =
+  ( Logic t
+  , TypeList dom
+  , Eq (t dom rng)
+  , Show (t dom rng)
+  , Typeable dom
+  , Typeable rng
+  , All HasSpec dom
+  , HasSpec rng
+  )
+
+data Term a where
+  App ::
+    forall t dom rng.
+    AppRequires t dom rng =>
+    t dom rng ->
+    List Term dom ->
+    Term rng
+  Lit :: (Typeable a, Eq a, Show a) => a -> Term a
+  V :: HasSpec a => Var a -> Term a
+
+instance Eq (Term a) where
+  V x == V x' = x == x'
+  Lit a == Lit b = a == b
+  App (w1 :: x1) (ts :: List Term dom1) == App (w2 :: x2) (ts' :: List Term dom2) =
+    case (eqT @dom1 @dom2, eqT @x1 @x2) of
+      (Just Refl, Just Refl) ->
+        w1 == w2
+          && sameTerms ts ts'
+      _ -> False
+  _ == _ = False
+
+-- | If the list is composed solely of literals, apply the function to get a value
+applyFunSym ::
+  forall t ds rng.
+  (Typeable rng, Eq rng, Show rng, Semantics t) => FunTy ds rng -> List Term ds -> Maybe rng
+applyFunSym f Nil = Just f
+applyFunSym f (Lit x :> xs) = applyFunSym @t (f x) xs
+applyFunSym _ _ = Nothing
+
+reducesToLit :: Term a -> Maybe a
+reducesToLit (Lit n) = Just n
+reducesToLit (V _) = Nothing
+reducesToLit (App (f :: t ds r) xs) = applyFunSym @t (semantics f) xs
+
+-- How to compare the args of two applications for equality
+sameTerms :: All HasSpec as => List Term as -> List Term as -> Bool
+sameTerms Nil Nil = True
+sameTerms (x :> xs) (y :> ys) = x == y && sameTerms xs ys
+
+-- ===========================================
+-- Function Symbol Classes
+-- ===========================================
+
+-- | Syntactic operations are ones that have to do with the structure and appearence of the type.
+class Syntax (t :: [Type] -> Type -> Type) where
+  inFix :: forall dom rng. t dom rng -> Bool
+  inFix _ = False
+  name :: forall dom rng. t dom rng -> String
+
+-- | Semantic operations are ones that give the function symbol, meaning as a function.
+--   I.e. how to apply the function to a list of arguments and return a value,
+--   or to apply meaning preserving rewrite rules.
+class Syntax t => Semantics (t :: [Type] -> Type -> Type) where
+  semantics :: forall d r. t d r -> FunTy d r -- e.g. FunTy '[a,Int] Bool == a -> Int -> Bool
+  rewriteRules ::
+    forall ds rng.
+    (TypeList ds, Typeable ds, HasSpec rng, All HasSpec ds) =>
+    t ds rng -> List Term ds -> Evidence (Typeable rng, Eq rng, Show rng) -> Maybe (Term rng)
+  rewriteRules t l Evidence = Lit <$> (applyFunSym @t (semantics t) l)
+
+-- | Logical operations are one that support reasoning about how a function symbol
+--   relates to logical properties, that we call Spec's
+class (Typeable t, Syntax t, Semantics t) => Logic (t :: [Type] -> Type -> Type) where
```

Comment:
We could get rid of these classes with a single universe and have only the functions defined on the universe instead.